src/software/new_geom/polygon.cpp

```diff
@@ -2,9 +2,33 @@
 
 #include <unordered_set>
 
-Polygon::Polygon(const std::vector<Point>& points) : points_(points) {}
+Polygon::Polygon(const std::vector<Point>& points)
+    : points_(points), segments_(initSegments(points_))
+{
+    // we pre-compute the segments_ in the constructor to improve performance
+}
 
-Polygon::Polygon(const std::initializer_list<Point>& points) : points_(points) {}
+Polygon::Polygon(const std::initializer_list<Point>& points)
+    : Polygon(std::vector(points))
+{
+}
+
+std::vector<Segment> Polygon::initSegments(std::vector<Point> points)
+{
+    std::vector<Segment> segments;
+    for (unsigned i = 0; i < points.size(); i++)
+    {
+        // add a segment between consecutive points, but wrap index
+        // to draw a segment from the last point to first point.
+        segments.emplace_back(Segment{points[i], points[(i + 1) % points.size()]});
+    }
+    return segments;
+}
+
+bool Polygon::invariantsHold() const
+{
+    return initSegments(points_) == segments_;
+}
```

Comment:
I'm a bit confused, why did we need this?

Reply:
bumping this

Reply:
fine we don't...